Expose push_subseq method on TreeBuilder

raphlinus opened the pull request:
It's cleaner to have this method on TreeBuilder than Tree. It's also a
useful an important method; the general template for any incremental
algorithm is to use a TreeBuilder to assemble its result, and I think we
should encourage its use.

- [ ] I have responded to reviews and made changes where appropriate.
- [x] I have tested the code with `cargo test --all` / `./rust/run_all_checks`.
- [x] I have updated comments / documentation related to the changes I made.
- [x] I have rebased my PR branch onto xi-editor/master.

raphlinus commented:
One question for discussion: I preserved the name "subseq", which was chosen at the time I think largely to accommodate the needs of span builders, but I think perhaps `push_slice` might be a cleaner name, and more consistent with `push_leaf_slice` already exists.

cmyr commented on rust/rope/src/tree.rs:
```diff
@@ -569,6 +541,35 @@ impl<N: NodeInfo> TreeBuilder<N> {
         }
     }
 
+    /// Push a subsequence of a rope.
```
This should be clearer about what the `iv` argument is.

cmyr reviewed (commented):
I do think this is a good time to get the naming and docs right, here.

I think `push_slice` or `push_interval` probably makes sense. I would also clarify the behaviour in the docs, since I think the arguments are not self-explanatory.

raphlinus commented:
Changed the name and added some docstring.

There's a bunch of potential cleanup to do here. For one, I'm not sure `Interval` is pulling its weight, it should probably be `Range`.

Aside from cleanup, for use as an immutable vector in Druid there's functionality we should add, especially in-place mutation. But overall I want to be conscious of time. If there's a person interested in taking on xi-rope improvements, it's something I could mentor.

cmyr reviewed (approved):
Docs are good, thanks!
